content/post/2019-11-01-internal-functions.Rmarkdown

```diff
@@ -0,0 +1,185 @@
+---
+title: Internal functions in R packages
+date: '2019-11-01'
+slug: internal functions
+tags:
+  - package development
+---
+
+An R package can be viewed as a [set of functions](https://github.com/ropensci/software-review/issues/350#issue-518124603), of which only a part are exposed to the user. In this blog post we shall concentrate of the functions that are not exposed to the user, so called internal functions: what are they, how does one handle them in one's own package, and how can one explore them?
+
+## Internal functions 101
+
+### What is an internal function? 
+
+It's a function that lives in your package, but that isn't surfaced to the user. You could also call it unexported function or helper function; as opposed to [exported functions](https://r-pkgs.org/namespace.html#exports) and user-facing functions.
+
+For instance, in the usethis package there's a `base_and_recommended()` function that is not exported.
+
+```{r examples, error = TRUE}
+# doesn't work
+library("usethis")
+base_and_recommended()
+usethis::base_and_recommended()
+# works
+usethis:::base_and_recommended()
+```
+
+### Why not export all functions?
+
+There are at least these two reasons:
+
+* In a package you want to provide your user an API that is useful and stable. You can vouch for a few functions, that work well, serve the package main goals, are documented enough, and that you'd only change [with great care](https://devguide.ropensci.org/evolution.html) [if need be](https://ropensci.org/blog/2019/04/30/qualtrics-relaunch/). If your package users rely of an internal function that you decide to ditch when re-factoring code, they won't be happy, so only export what you want to maintain.
```

Comment:
I would remove "that work well". The internal functions are supposed to work well as well, no? :)